### PR TITLE
Use bison for ilasm parser generator

### LINKED ISSUE #4776: Generate asmparse.c using a standard bison/yacc
asmparse.c (for ilasm) is currently prebuilt using an internal version of yacc. 
This also caused many compilations warnings on x-platforms.

### PR DESCRIPTION
**Pull request by hez2010**
Supersedes #89704. Fixes #4776.

Bison for Windows: https://github.com/lexxmark/winflexbison
For other platforms: https://www.gnu.org/software/bison

To generate the `asmparse.cpp`: `yacc asmparse.y -o asmparse.cpp` (or `win_bison asmparse.y -o asmparse.cpp`)

/cc: @TIHan @JulieLeeMSFT

### CODE REVIEW

File: src/coreclr/ilasm/asmparse.y [hunk 1]
```diff
@@ -42,13 +42,13 @@
 };
 
         /* These are returned by the LEXER and have values */
-%token ERROR_ BAD_COMMENT_ BAD_LITERAL_                         /* bad strings,    */
```
Comment (hez2010) [inline src/coreclr/ilasm/asmparse.y:45]: This has been defined multiple times, so remove it here.

File: src/coreclr/ilasm/asmparse.y [hunk 2]
```diff
@@ -96,7 +96,7 @@
 %token _CLASS _NAMESPACE _METHOD _FIELD _DATA _THIS _BASE _NESTER
 %token _EMITBYTE _TRY _MAXSTACK _LOCALS _ENTRYPOINT _ZEROINIT
 %token _EVENT _ADDON _REMOVEON _FIRE _OTHER
-%token _PROPERTY _SET _GET DEFAULT_
```
Comment (hez2010) [inline src/coreclr/ilasm/asmparse.y:99]: This has been defined multiple times, so remove it here.

File: src/coreclr/ilasm/asmparse.y [hunk 3]
```diff
@@ -42,13 +42,13 @@
...
 
         /* These are returned by the LEXER and have values */
-%token ERROR_ BAD_COMMENT_ BAD_LITERAL_                         /* bad strings,    */
+%token BAD_COMMENT_ BAD_LITERAL_                         /* bad strings,    */
 %token <string>  ID             /* testing343 */
 %token <string>  DOTTEDNAME     /* System.Object */
 %token <binstr>  QSTRING        /* "Hello World\n" */
 %token <string>  SQSTRING       /* 'Hello World\n' */
-%token <int32>   INT32          /* 3425 0x34FA  0352  */
-%token <int64>   INT64          /* 342534523534534      0x34FA434644554 */
```
Comment (hez2010) [inline src/coreclr/ilasm/asmparse.y:51]: They are conflicting with `INT32` and `INT64` typedef, so rename them.

File: src/coreclr/ilasm/asmparse.y [hunk 4]
```diff
@@ -42,13 +42,13 @@
...
+%token <int32>   INT32_V        /* 3425 0x34FA  0352  */
+%token <int64>   INT64_V        /* 342534523534534      0x34FA434644554 */
```
Comment (hez2010) [inline src/coreclr/ilasm/asmparse.y:51]: They are conflicting with `INT32` and `INT64` typedef, so rename them.

**Comment by TIHan** [comment]
Thank you @hez2010 ! Will review this soon.

**Review by TIHan — COMMENTED** [review]
So far this looks good. Going to look at it a bit deeper.

File: src/coreclr/ilasm/asmparse.y (hunk 1 above)
Comment (TIHan) [reply src/coreclr/ilasm/asmparse.y:45]: Makes sense to me.

File: src/coreclr/ilasm/asmparse.y (hunk 2 above)
Comment (TIHan) [reply src/coreclr/ilasm/asmparse.y:99]: Yup, makes sense.

File: src/coreclr/ilasm/asmparse.y (hunk 4 above)
Comment (TIHan) [reply src/coreclr/ilasm/asmparse.y:51]: Yea, IIRC, the previous PR also had to do this.

**Comment by AaronRobinsonMSFT** [comment]
> I have deleted the asmparse.grammar file as it's no longer being maintained since 2021.

I question this is no longer being maintained. What IL grammar changes have we introduced since 2021 that havn't made it in?

**Comment by hez2010** [comment]
> What IL grammar changes have we introduced since 2021 that haven't made it in?

For example, byreflike on generics.

**Comment by AaronRobinsonMSFT** [comment]
> > What IL grammar changes have we introduced since 2021 that haven't made it in?
> 
> For example, byreflike on generics.

I wrote that feature and it was updated. See `byreflike` under `typarAttrib`.

**Comment by hez2010** [comment]
> I wrote that feature and it was updated. See `byreflike` under `typarAttrib`.

Okay I see, sorry for missed that. I can bring it back but note that there's no public tool that can auto generate the asmparser.grammar file now.

**Review by TIHan — APPROVED** [review]
Thanks for your work @hez2010 ! Glad to see this be done.

I'll make a follow-up PR that adds a small README.md on how to generate the parser.

**Pull request merged**